Retriever and lvm update for multimodal rag on videos

tileintel opened the pull request:
## Description

This PR introduces multimodal retriever microservice and provides update on LVM microservice. They are required for Multimodal RAG on Videos application. This allow users to retrieve text and its associated metadata as multimodal data from Redis vector store. The update on LVM microservice allows LVM accepts as input SearchedMultimodalDoc (additionally to LVMDoc). Such SearchedMultimodalDoc is the output of multimodal retriever microservice. 

## Issues

RFC: [https://github.com/opea-project/docs/pull/49/files](https://github.com/opea-project/docs/pull/49/files)
Issue: [https://github.com/opea-project/GenAIExamples/issues/358](https://github.com/opea-project/GenAIExamples/issues/358)

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds new functionality) Introducting multimodal retriever microservice
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [ x] Others (enhancement, documentation, validation, etc.) Enhancing LVM microservice

## Dependencies

docarray[full]
fastapi
langchain_community
opentelemetry-api
opentelemetry-exporter-otlp
opentelemetry-sdk
prometheus-fastapi-instrumentator
redis
shortuuid
uvicorn
transformers
[llava-hf/llava-1.5-7b-hf](https://huggingface.co/llava-hf/llava-1.5-7b-hf)
[BridgeTower](https://huggingface.co/BridgeTower/bridgetower-large-itm-mlm-gaudi)
langchain-core

## Tests

We have provided 2 tests for these microservice.
- tests/test_lvms_llava.sh. We provided 2 additional tests within validate_microservice() for new request type of LVMSearchedMultimodalDoc. 
- tests/test_retrievers_multimodal_langchain_redis.sh. This tests the introduced multimodal retriever microservice.

tileintel commented:
@lvliang-intel @XuhuiRen This is the last PR for GenAIComps on Multimodal RAG on Videos. Would you please review this? We will need this PR for GenAIExamples. 
Thanks

lkk12014402 commented on comps/retrievers/multimodal_langchain/redis/multimodal_config.py:
```diff
@@ -0,0 +1,82 @@
+
+REDIS_SCHEMA = os.getenv("REDIS_SCHEMA", "redis_schema.yml")
+schema_path = os.path.join(parent_dir, REDIS_SCHEMA)
+INDEX_SCHEMA = schema_path
```
where use this variable?

tileintel replied:
@lkk12014402  Thank for your feedback. Yes, these variables are not used by retriever microservice. I have removed REDIS_SCHEMA, INDEX_SCHEMA, and NUM_RETRIEVED_RESULTS. Thanks

lkk12014402 commented:
hi, @tileintel @lvliang-intel I am confused about the directory structure,  what about using this directory structure `comps/retrievers/langchain/redis` and putting the multi-modal implementations into `comps/retrievers/langchain/redis`?

@XuhuiRen  do you have some comments for this?

tileintel commented:
> hi, @tileintel @lvliang-intel I am confused about the directory structure, what about using this directory structure `comps/retrievers/langchain/redis` and putting the multi-modal implementations into `comps/retrievers/langchain/redis`?
> 
> @XuhuiRen do you have some comments for this?

Dear @lkk12014402. 
Thank you for your feedback. At first, I totally agree that we can put this implementation in langchain/redis. However, we prefer having separate folder as multimodal_langchain/redis due to the following reasons:
1. Langchain as you know is developed mainly for language modality only. Thus, in order to extend langchain to accept multimodalities, we actually had to extend some langchain original classes to new multimodal corresponding class. We developed those classes but not yet submit PR to langchain open-source repo yet. Thus, having separate folders helps us to switch back to original langchain implementation easily in future if langchain community decides to accept multimodalities and have different implementation to what we have implemented.
2. You might not see any new implementation for langchain in this retriever microservice. However, there are quite a bit of them in another microservice like multimodal_embedding. And in other microservice implementation, we already have a folder structure like this, so we propose the similar folder structure here for consistency.
3. imho, langchain that accepts secondary modality like image is not original langchain anymore. It should be multimodal_langchain. And redis class here is an interface between langchain and Redis database. Similarly, new redis class (if existing) for multimodal_langchain should not be in `langchain/multimodal_redis`. 

Given our observations above, would you please reconsider this? I hope these can address your concern. 
Again, thank a lot for your insight feedback/question.

XuhuiRen commented:
> comps/retrievers/langchain/redis

I think it would be better to put the retriever under "comps/retrievers/langchain/". Currently, we category the type of the retriever according to their implementation framework. it is not reasonable to create a parallel folder with langchain and llamaindex.

XuhuiRen commented on comps/retrievers/multimodal_langchain/redis/README.md:
please revise the folder structure

XuhuiRen reviewed (changes requested):
Please revise the folder structure

tileintel commented:
> Please revise the folder structure

@XuhuiRen: Sure, I can revise the folder structure as you suggested. However, if we leave these new files in langchain/redis/, what files' names should I use for README, Dockerfile and docker_compose yml files? There will be two README files: one for retriever microservice and one for multimodal retriever microservice. Also, should I name dockerfile as Dockerfile_multimodal? Please advise. 
Given this, langchain/multimodal_redis is also not a good solution as well.
Thanks a lot

XuhuiRen commented:
> > Please revise the folder structure
> 
> @XuhuiRen: Sure, I can revise the folder structure as you suggested. However, if we leave these new files in langchain/redis/, what files' names should I use for README, Dockerfile and docker_compose yml files? There will be two README files: one for retriever microservice and one for multimodal retriever microservice. Also, should I name dockerfile as Dockerfile_multimodal? Please advise. Given this, langchain/multimodal_redis is also not a good solution as well. Thanks a lot

you can temperately create a redis-multimodal folder in the retriever/langchain folder. we can schedule a movement to rename the folder structure like base_redis, multimodal_redis in the next release

tileintel commented:
> > > Please revise the folder structure
> > 
> > 
> > @XuhuiRen: Sure, I can revise the folder structure as you suggested. However, if we leave these new files in langchain/redis/, what files' names should I use for README, Dockerfile and docker_compose yml files? There will be two README files: one for retriever microservice and one for multimodal retriever microservice. Also, should I name dockerfile as Dockerfile_multimodal? Please advise. Given this, langchain/multimodal_redis is also not a good solution as well. Thanks a lot
> 
> you can temperately create a redis-multimodal folder in the retriever/langchain folder. we can schedule a movement to rename the folder structure like base_redis, multimodal_redis in the next release

@XuhuiRen Thank for suggestions. I have revised folder structure as you suggested. In specific, all files are moved to comps/retrievers/langchain/redis_multimodal. 
I have updated README and test_retrievers_langchain_multimodal_redis.sh accordingly.

lkk12014402 commented:
> > > > Please revise the folder structure
> > > 
> > > 
> > > @XuhuiRen: Sure, I can revise the folder structure as you suggested. However, if we leave these new files in langchain/redis/, what files' names should I use for README, Dockerfile and docker_compose yml files? There will be two README files: one for retriever microservice and one for multimodal retriever microservice. Also, should I name dockerfile as Dockerfile_multimodal? Please advise. Given this, langchain/multimodal_redis is also not a good solution as well. Thanks a lot
> > 
> > 
> > you can temperately create a redis-multimodal folder in the retriever/langchain folder. we can schedule a movement to rename the folder structure like base_redis, multimodal_redis in the next release
> 
> @XuhuiRen Thank for suggestions. I have revised folder structure as you suggested. In specific, all files are moved to comps/retrievers/langchain/redis_multimodal. I have updated README and test_retrievers_langchain_multimodal_redis.sh accordingly. Would you please review this PR again and merge if there is no new feedback?
> 
> Thanks a lot

some uts are not passed, please check~

tileintel commented:
> > > > > Please revise the folder structure
> > > > 
> > > > 
> > > > @XuhuiRen: Sure, I can revise the folder structure as you suggested. However, if we leave these new files in langchain/redis/, what files' names should I use for README, Dockerfile and docker_compose yml files? There will be two README files: one for retriever microservice and one for multimodal retriever microservice. Also, should I name dockerfile as Dockerfile_multimodal? Please advise. Given this, langchain/multimodal_redis is also not a good solution as well. Thanks a lot
> > > 
> > > 
> > > you can temperately create a redis-multimodal folder in the retriever/langchain folder. we can schedule a movement to rename the folder structure like base_redis, multimodal_redis in the next release
> > 
> > 
> > @XuhuiRen Thank for suggestions. I have revised folder structure as you suggested. In specific, all files are moved to comps/retrievers/langchain/redis_multimodal. I have updated README and test_retrievers_langchain_multimodal_redis.sh accordingly. Would you please review this PR again and merge if there is no new feedback?
> > Thanks a lot
> 
> some uts are not passed, please check~

One UT failed because of something happened in Circle CI in which it cannot clear the previously started dockers and some ports were not available for start some microservice for that UT. I have changed the port number in the tests. It seems working now.
@lkk12014402 @XuhuiRen @lvliang-intel Since all feedbacks are resolved and all the tests work, can you please merge this PR?
Thanks a lot

tileintel commented:
@XuhuiRen @lkk12014402 @lvliang-intel  Since all tests have passed and there is no more feedbacks, can anyone please merge this PR? Thanks a lot for helping this.
